src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeEditor.cs

```diff
@@ -0,0 +1,226 @@
+
+                // Get the original source document again, from the latest forked solution.
+                var sourceDocument = solutionWithNewDocument.GetDocument(SemanticDocument.Document.Id);
+
+                // update source document to add partial modifiers to type chain
+                // and/or remove type declaration from original source document.
+                var solutionWithBothDocumentsUpdated = await RemoveTypeFromSourceDocumentAsync(
+                      sourceDocument, State.TypeNode, CancellationToken).ConfigureAwait(false);
+
+                return SpecializedCollections.SingletonEnumerable(new ApplyChangesOperation(solutionWithBothDocumentsUpdated));
+            }
+
+            /// <summary>
+            /// Forks the source document, keeps required type, namespace containers
+            /// and adds it the solution.
+            /// </summary>
+            /// <param name="sourceDocument">original document</param>
+            /// <param name="newDocumentName">name of the new document to be added</param>
+            /// <param name="newDocumentId">id for the new document to be added</param>
+            /// <param name="typeNode">type to move from original document to new document</param>
+            /// <param name="cancellationToken">a cancellation token</param>
+            /// <returns>the new solution which contains a new document with the type being moved</returns>
+            private async Task<Solution> AddNewDocumentWithSingleTypeDeclarationAndImportsAsync(
+                SemanticDocument sourceDocument,
+                string newDocumentName,
+                DocumentId newDocumentId,
+                TTypeDeclarationSyntax typeNode,
+                CancellationToken cancellationToken)
+            {
+                Debug.Assert(sourceDocument.Document.Name != newDocumentName,
+                             $"New document name is same as old document name:{newDocumentName}");
+
+                var root = sourceDocument.Root;
+                var projectToBeUpdated = sourceDocument.Document.Project;
+                var documentEditor = await DocumentEditor.CreateAsync(sourceDocument.Document, cancellationToken).ConfigureAwait(false);
+
+                AddPartialModifiersToTypeChain(documentEditor, typeNode);
+
+                // remove things that are not being moved, from the forked document.
+                var membersToRemove = GetMembersToRemove(root, typeNode);
+                foreach (var member in membersToRemove)
+                {
+                    documentEditor.RemoveNode(member, SyntaxRemoveOptions.KeepNoTrivia);
+                }
+
+                var modifiedRoot = documentEditor.GetChangedRoot();
+
+                // add an empty document to solution, so that we'll have options from the right context.
+                var solutionWithNewDocument = projectToBeUpdated.Solution.AddDocument(newDocumentId, newDocumentName, text: string.Empty);
```

Comment:
what happens if there's already a file on disk with that name, but it's not in the project?